src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -109,22 +132,28 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
 			if (cacheItem.isExpired())
 				refreshCacheItemTask(cacheItem); // Refresh the cacheItem in a separate thread
 		} else { // Item is not in cache.
-			ProductService ps = new ProductService(dto);
-			Locale locale = Locale.forLanguageTag(dto.getLocale());
-			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
-				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
-				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
-					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
-			} else   // Requested locale is not supported and does not match any supported locales
+			cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
+			if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
 				cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 		}
+
 		return cacheItem;
 	}
 
 	private void doLocaleMatching() {
-		Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new ProductService(dto).getSupportedLocales(), Locale.forLanguageTag(dto.getLocale()));
-		if (matchedLocale != null) // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+		dto.setLocale(LocaleUtility.fmtToMappedLocale(dto.getLocale()).toLanguageTag());
+
+		ProductService ps = new ProductService(dto);
+
+		//Refresh the cache of supported locales as needed in a separate thread (non-blocking).
+		ps.refreshSupportedLocalesTask();
+
+		//Match against list of supported locales that is already in the cache
+		Set<Locale> supportedLocales = LocaleUtility.langTagtoLocaleSet(ps.getCachedSupportedLocales());
+		Locale matchedLocale = LocaleUtility.pickupLocaleFromList(supportedLocales, Locale.forLanguageTag(dto.getLocale()));
+		if (matchedLocale != null) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
 			dto.setLocale(matchedLocale.toLanguageTag());
+		}
```

Comment:
Looks pretty much the same to me. Based on the flow you described
1. ComponentService will get supported locales from product service
2. The product service returns an empty set if cache of locales is not found.
3. The Product Service will refresh the cache if needed (Note: must be in a separate thread).
4. If supported locales set isn't null/empty, then pick up the locale from them.

# 1 This PR already does it when ComponentService.doLocaleMatching calls ps.getCachedSupportedLocales(). 
# 2 In this PR,  ProductService.getCachedSupportedLocales() in # 1 returns an empty set if not found in cache. 
# 3 Sure, I can call ProductService.refreshSupportedLocalesTask inside ProductService.getCachedSupportedLocales instead of here.
# 4 This Pr does it when ComponentService.doLocaleMatching sets dto.locale with the matchedlocale: `dto.setLocale(matchedLocale.toLanguageTag());`

The only added logic to the flow is ComponentService.proceed, which is only called when message bundle is not yet in cache or is expired. This is for performance enhancement.  ComponentService.proceed calls ps.getCachedSupportedLocales(dataSource) to check if the requested locale (dto.getLocale()) is supported in that specific data source. This is so that if it is not supported in that data source, it won't even try to fetch the message bundle. This is important especially if that invalid fetch is to a remote VIP service which will affect performance.